[New offload][llc] Pass -relocation-model=pic option to llc when building shared libraries

asudarsa opened the pull request:
When user is building a shared library using '-shared' option, the llc step run after wrapping device code needs to be run with '-relocation-model=pic' option. This PR adds the above option for the new offloading model.
One other change has been added to address test failures resulting from adding a new SYCL E2E test:
- Wrapper logic needs to use the 'arch' to specify the target and not use the entire triple. This is required for successful parsing by SYCL runtime.

Thanks

mdtoguchi commented on clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp:
```diff
@@ -814,6 +814,13 @@ static Expected<StringRef> runCompile(StringRef &InputFile,
   SmallVector<StringRef, 8> CmdArgs;
   CmdArgs.push_back(*LLCPath);
+  // Checking for '-shared' linker option
+  for (const opt::Arg *Arg : Args) {
+    if (Arg->getOption().matches(OPT_shared)) {
```
Instead of looping through the args, maybe `Args.hasArg(OPT_shared)` would be better

asudarsa replied:
Ah.. Good point. Thanks Mike. Will address.

asudarsa replied:
Done

asudarsa commented on sycl/test/regression/joint_group_conflict_new_offload.cpp:
```diff
@@ -0,0 +1,28 @@
+// RUN: %clangxx -fsycl -fsycl-targets=%sycl_triple --offload-new-driver -fPIC -DCASE1 %s -c -o %t.1.o
+// RUN: %clangxx -fsycl -fsycl-targets=%sycl_triple --offload-new-driver -fPIC -DCASE2 %s -c -o %t.2.o
+// RUN: %clangxx -fsycl -fsycl-targets=%sycl_triple --offload-new-driver -shared %t.1.o %t.2.o -o %t.so
```
Thanks for feedback Andrei. Will address.

asudarsa replied:
I will also try to add a driver test to test for the option to be passed in correctly. Thanks again

asudarsa replied:
Added relevant tests. Thanks

asudarsa commented on clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp:
```diff
@@ -767,8 +767,13 @@ Expected<StringRef> wrapSYCLBinariesFromFile(StringRef InputFile,
         inconvertibleErrorCode(),
         "can't wrap SYCL image. -triple argument is missed.");
 
+  // SYCL runtime currently works for spir64 target triple and not for
```
This change is not directly in the scope of this PR. But it fixes an issue that was exposed by running E2E SYCL test.

Thanks

mdtoguchi commented on clang/test/Driver/linker-wrapper-sycl.cpp:
```diff
@@ -10,3 +10,7 @@
 // CHK-CMDS-NEXT: "{{.*}}/ld" -- HOST_LINKER_FLAGS -dynamic-linker HOST_DYN_LIB -o a.out [[LLCOUT]].o HOST_LIB_PATH HOST_STAT_LIB {{.*}}test-sycl.o
+
+/// check for PIC for device wrap compilation when using -shared
+// RUN: clang-linker-wrapper -sycl-device-library-location=%S/Inputs -sycl-device-libraries=libsycl-crt.new.o,libsycl-complex.new.o -sycl-post-link-options="SYCL_POST_LINK_OPTIONS" -llvm-spirv-options="LLVM_SPIRV_OPTIONS" "--host-triple=x86_64-unknown-linux-gnu" "--triple=spir64" "--linker-path=/usr/bin/ld" -shared "--" HOST_LINKER_FLAGS "-dynamic-linker" HOST_DYN_LIB "-o" "a.out" HOST_LIB_PATH H ...
```
Are all of the additional input options necessary when just checking for the LLC call?

asudarsa replied:
I just wanted to keep it simple and reuse what was working already. Hope it's ok.

Thanks

mdtoguchi replied:
Fine by me.  Thanks!

asudarsa commented:
CI testing revealed two issues:
1. clang-offload-packager and clang-offload-linker are not being deployed
- This was missed in the initial PR which added these tools to CMakeLists.txt. Fixed here.
2. AOT builds fail under new-offload-driver
- This is expected as we have not yet added AOT support for new offload model. Modified the tests according to latest status.

Thanks

aelovikov-intel commented on sycl/test-e2e/SharedLib/use_when_link_new_offload.cpp:
```diff
@@ -0,0 +1,64 @@
+// RUN: %clangxx -fsycl -fsycl-targets=spir64 --offload-new-driver -L%T %s -o %t.out -l%basename_t -Wl,-rpath=%T
+// RUN: %{run} %t.out
+
+#include <sycl/detail/core.hpp>
```
Is this the same as the old test? Can we just `#include` it here?

asudarsa replied:
This is a good observation. It makes test maintenance also a little easier. Thanks. I have changed the test appropriately. Thanks

asudarsa commented on sycl/test-e2e/SharedLib/use_when_link_new_offload.cpp:
```diff
@@ -1,4 +1,4 @@
-// REQUIRES: linux
+// REQUIRES: opencl, cpu, linux
```
This test is expected to check if the shared library support works with new offload model. An extensive testing for the new offloading model for all targets will be accomplished later.

Thanks
